File: daily_problems/2026/04/0430/personal_submission/cf106507h_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int M = 1e6 + 5;
	vector<int> pr(M);
	iota(pr.begin(), pr.end(), 0);

	for (int i = 2; i < M; i ++) {
		if (pr[i] == i) {
			for (int j = i; j < M; j += i) {
				pr[j] = i;
			}
		}
	}

	int n, k;
	cin >> n >> k;

	vector<int> nums(n);
	for (auto &x: nums) cin >> x;

	vector<int> cnt(M, 0);

	for (auto x: nums) {
		while (x > 1) {
			cnt[pr[x]] ++;
			x /= pr[x];
		}
	}

	vector<int> dp(k + 1, 1);

	for (int i = 0; i < M; i ++) {
		if (cnt[i] >= n) {
			vector<int> pws(20, 0);

			for (auto &x: nums) {
				int pw = 0;
				while (x % i == 0) {
					x /= i;
					pw ++;
				}
				pws[pw] ++;
			}

			vector<pair<int, int>> transitions;
			int val = 1;

			for (int a = 1; a <= cnt[i] / n; a ++) {
				val *= i;
				int cost = 0;
				for (int b = 0; b < a; b ++)
					cost += (a - b) * pws[b];
				transitions.emplace_back(cost, val);
			}

			for (int i = k; i >= 0; i --) {
				int dp_val = dp[i];
				for (auto &[cost, val]: transitions) {
					if (i >= cost) {
						dp_val = max(dp_val, dp[i - cost] * val);
					}
				}
				dp[i] = dp_val;
			}
		}
	}

	for (int i = 1; i <= k; i ++) cout << dp[i] << ' ';

	return 0;
}

File: daily_problems/2026/04/0430/personal_submission/cf106507h_yawn_sean.py
# Submission link: https://codeforces.com/gym/106507/submission/372917868
def main():
    M = 10 ** 6 + 5
    pr = list(range(M))
    
    for i in range(2, M):
        if pr[i] == i:
            for j in range(i, M, i):
                pr[j] = i
    
    n, k = MII()
    nums = LII()
    
    cnt = [0] * M
    
    for x in nums:
        while x > 1:
            cnt[pr[x]] += 1
            x //= pr[x]
    
    dp = [1] * (k + 1)
    
    for i in range(M):
        if cnt[i] >= n:
            
            pws = [0] * 20
            for x in nums:
                pw = 0
                while x % i == 0:
                    x //= i
                    pw += 1
                pws[pw] += 1
    
            transitions = []
            val = 1
            
            for a in range(cnt[i] // n + 1):
                cost = 0
                for b in range(a):
                    cost += (a - b) * pws[b]
                transitions.append((cost, val))
                val *= i
            
            for i in range(k, -1, -1):
                dp_val = dp[i]
                for cost, val in transitions:
                    if i >= cost:
                        dp_val = fmax(dp_val, dp[i - cost] * val)
                dp[i] = dp_val
    
    print(' '.join(map(str, dp[1:])))